lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -265,42 +268,256 @@ class ConvertAten_IndexPutImplOp
-    SmallVector<int64_t> expandedIndexTensorSizes{indexTensorSize, 1};
-    ValueTensorType expandedIndexTensorType =
-        ValueTensorType::get(context, llvm::ArrayRef(expandedIndexTensorSizes),
-                             indexTensorType.getDtype());
+    Value torchCstZero = rewriter.create<Torch::ConstantIntOp>(
+        loc, rewriter.getI64IntegerAttr(0));
     Value torchCstOne = rewriter.create<Torch::ConstantIntOp>(
         loc, rewriter.getI64IntegerAttr(1));
-    Value expandedIndexTensor = rewriter.create<AtenUnsqueezeOp>(
-        loc, expandedIndexTensorType, indexTensor, torchCstOne);
+    Value torchCstNone = rewriter.create<Torch::ConstantNoneOp>(loc);
+    Value indexTensor, flattenedValuesTensorLastDim;
+    BaseTensorType indexTensorType;
+    bool needToFlattenValuesTensor = false;
+    SmallVector<int64_t> flattenedValuesTensorSize;
+    if (indicesList.size() == 1) {
+      indexTensor = indicesList[0];
+      if (indexTensor.getType().isa<Torch::NoneType>())
+        return rewriter.notifyMatchFailure(op,
+                                           "Index tensor must not be None.");
+      indexTensorType = indexTensor.getType().cast<BaseTensorType>();
+      std::optional<unsigned> indexTensorRank = getTensorRank(indexTensor);
+      if (!indexTensorRank)
+        return rewriter.notifyMatchFailure(
+            op, "expected index tensor to have a rank");
+
+      if (*indexTensorRank == 2) {
+        SmallVector<int64_t> indexTensorShape(indexTensorType.getSizes());
+        if (indexTensorShape[0] == 1) {
+          ValueTensorType flattenedIndexTensorType = ValueTensorType::get(
+              context, llvm::ArrayRef({indexTensorShape[1]}),
+              indexTensorType.getDtype());
+          indexTensor = rewriter.create<AtenFlattenUsingIntsOp>(
+              loc, flattenedIndexTensorType, indexTensor, torchCstZero,
+              torchCstOne);
+        } else {
+          return rewriter.notifyMatchFailure(
+              op, "unimplemented: index tensor with rank 2 is supported only "
+                  "when its first dimension is equal to 1");
+        }
+      } else if (*indexTensorRank > 2) {
+        return rewriter.notifyMatchFailure(
+            op, "unimplemented: index tensor with rank > 2 is not supported");
+      }
+
+      indexTensorType = indexTensor.getType().cast<BaseTensorType>();
+      assert(indexTensorType.getSizes().size() == 1 &&
+             "expected index tensor to be of rank 1");
+      int64_t indexTensorSize = indexTensorType.getSizes()[0];
+      SmallVector<int64_t> expandedIndexTensorSizes{indexTensorSize, 1};
+      ValueTensorType expandedIndexTensorType = ValueTensorType::get(
+          context, llvm::ArrayRef(expandedIndexTensorSizes),
+          indexTensorType.getDtype());
+      indexTensor = rewriter.create<AtenUnsqueezeOp>(
+          loc, expandedIndexTensorType, indexTensor, torchCstOne);
+    } else if (indicesList.size() > 1 && indicesList.size() == 4) {
```

Comment:
This is too restrictive. I don't think it would require much more to have full support  of 2 index tensors. We should have a helper function that computes the broadcasted size of the index tensors (not sure if this already exists in torch-mlir) to then compute `final_shape`. Then it should be a matter of turning every `None_i` in the index list into a `torch.arange(0, input.size(i)).reshape(1, ..., input.size(i), ..., 1, 1).flatten()`.

Reply:
Hi @ramiro050, could you please review this patch now?

Reply:
Sure! I'm a bit busy right now, but I will for sure review it tomorrow.

Reply:
Thanks!